### PR TITLE
feat(backend): access token validation

### PR DESCRIPTION
**Pull request by Mogge**
### Issues
- relates #144

### Todo

- [X] None

### CODE REVIEW

File: frontend/src/pages/index/+Page.vue [hunk 1]
```diff
@@ -1,7 +1,36 @@
...
+      </v-col>
+    </v-row>
+  </DefaultLayout>
 </template>
 
 <script lang="ts" setup>
+import { useQuery } from '@vue/apollo-composable'
+
 import DefaultLayout from '#layouts/DefaultLayout.vue'
+// eslint-disable-next-line import/no-relative-parent-imports
```
Comment (Mogge) [inline frontend/src/pages/index/+Page.vue:22]: any ideas how to fix this?

File: frontend/src/pages/index/Page.test.ts [hunk 2]
```diff
@@ -20,4 +40,39 @@ describe('IndexPage', () => {
...
+      it('calls the API', () => {
+        expect(getRoomQueryMock).toBeCalled()
+      })
+
+      it('redirects to url', () => {
+        expect(global.window.location.href).toBe('http://some.url/')
+      })
+    })
+
+    describe.skip('with error', () => {
```
Comment (Mogge) [inline frontend/src/pages/index/Page.test.ts:62]: I do not know why this is not working :(

File: backend/src/server/server.ts [hunk 3]
```diff
@@ -3,18 +3,23 @@ import { startStandaloneServer } from '@apollo/server/standalone'
...
-import logger from './logger'
+import { Context, getContextToken, GetConextToken } from './context'
+// import logger from './logger'
 
 export const createServer = async (): Promise<ApolloServer> => {
-  return new ApolloServer({
+  return new ApolloServer<Context>({
     schema: await schema(),
-    plugins: [logger],
+    // plugins: [logger],
```
Comment (Mogge) [inline backend/src/server/server.ts:12]: I created an issue for this: #463

File: frontend/src/pages/index/+Page.vue [hunk 4]
```diff
@@ -1,7 +1,39 @@
...
+const apolloClient = inject<ApolloClient<InMemoryCache>>(DefaultApolloClient)
+
+const enterRoom = async () => {
+  try {
+    const result = await apolloClient?.query({ query: getRoomQuery, fetchPolicy: 'network-only' })
+    window.location.href = result?.data?.getRoom
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  } catch (error: any) {
+    // eslint-disable-next-line no-console
+    console.log('auth error', error.message ? error.message : error)
```
Comment (Mogge) [inline frontend/src/pages/index/+Page.vue:36]: any ideas how to do this better?

File: backend/src/server/context.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export type Context = {
+  token?: string
+}
+
+export type GetConextToken = (authorization: string | undefined) => string | undefined
+
+export const getContextToken: GetConextToken = (authorization) => {
+  return authorization ? authorization.replace(/^[Bb]earer */, '') : undefined
+}
```
Comment (ulfgebhardt) [inline backend/src/server/context.ts:9]:
```suggestion
export type GetContextToken = (authorization: string | undefined) => string | undefined

export const getContextToken: GetContextToken = (authorization) => {
  return authorization ? authorization.replace(/^[Bb]earer */, '') : undefined
}
```

File: backend/src/server/server.ts [hunk 6]
```diff
@@ -3,18 +3,23 @@ import { startStandaloneServer } from '@apollo/server/standalone'
 
 import { schema } from '#graphql/schema'
 
-import logger from './logger'
+import { Context, getContextToken, GetConextToken } from './context'
```
Comment (ulfgebhardt) [inline backend/src/server/server.ts:6]:
```suggestion
import { Context, getContextToken, GetContextToken } from './context'
```

File: backend/src/server/server.ts [hunk 7]
```diff
@@ -3,18 +3,23 @@ import { startStandaloneServer } from '@apollo/server/standalone'
...
-  return new ApolloServer({
+  return new ApolloServer<Context>({
     schema: await schema(),
-    plugins: [logger],
+    // plugins: [logger],
   })
 }
 
-export async function listen(port: number) {
+export async function listen(port: number, getToken: GetConextToken = getContextToken) {
```
Comment (ulfgebhardt) [inline backend/src/server/server.ts:16]:
```suggestion
export async function listen(port: number, getToken: GetContextToken = getContextToken) {
```

Comment (ulfgebhardt) [inline backend/src/server/server.ts:16]: Whats the reason you make this (`getToken`) a parameter? Are we expecting to make this exchangable?

File: frontend/src/pages/index/+Page.vue (hunk 4 above)
Comment (ulfgebhardt) [reply frontend/src/pages/index/+Page.vue:35]: https://v4.apollo.vuejs.org/guide-composable/error-handling.html#network-errors

This defines error handlers on link level and allows global error handling (e.g. popup or console)

**Comment by ulfgebhardt** [comment]
I can see it properly work - tokens are verified and access is denied when no valid token is available. Great work ❤️

File: backend/src/server/server.ts (hunk 7 above)
Comment (Mogge) [reply backend/src/server/server.ts:16]: I did this, that we can use a different context in the unit tests. When I did this, I did not know that you can pass `contextValue` to the `executeOperation` method.

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**